chorus_lib/src/transport/local.rs

```diff
@@ -5,52 +5,85 @@ use std::sync::Arc;
 
 use serde_json;
 
-use crate::core::{Portable, Transport};
+use std::marker::PhantomData;
+
+#[cfg(test)]
+use crate::LocationSet;
+
+use crate::core::{ChoreographyLocation, HList, Portable, Transport};
 use crate::utils::queue::BlockingQueue;
 
 type QueueMap = HashMap<String, HashMap<String, BlockingQueue<String>>>;
 
+/// A Transport channel used between multiple `Transport`s.
+pub struct LocalTransportChannel<L: crate::core::HList> {
+    /// The location set where the channel is defined on.
+    pub location_set: std::marker::PhantomData<L>,
+    queue_map: QueueMap,
+}
+
+impl<L: HList> LocalTransportChannel<L> {
+    /// Creates a `LocalTransportChannel`.
+    pub fn new() -> Arc<LocalTransportChannel<L>> {
```

Comment:
Three points:

First, it may be better to make this constructor private and define a wrapper on `LocalTransport` itself. This way, users don't have to import `LocalTransportChannel`.

```rust
let channel = LocalTransport::channel(...);
let transport = LocalTransport::new(Target, channel);
```

Second, I actually really like the builder pattern used in the HTTP transport, and I wonder if we can do something similar here.

```rust
let channel = LocalTransport::channel().with(Alice).with(Bob);
let alice_transport = LocalTransport::new(Alice, channel);
```

Third, if we change the `queue_map` field to `Arc<QueueMap>` and derive `Clone` on `LocalTransport`, can we just return `LocalTransportChannel`? I might have said something else, but now that I think about it, it might not be a good idea to expose the internal data structure (`Arc`) to the users. 

What do you think of these?

Reply:
I'm personally not a fan of the builder pattern for `LocalTransport`. There's no extra information assigned to each `Choreography` and it just feels extra. And on top of that, you have two lines for creating a transport instead of the one liner (`LocalTransportChannel::<LocationSet!(Bob, Alice)>::new()`)

Reply:
But if you want to have it anyways, sure! 


For your first point: what would be the argument of `LocalTransport::channel`? You mentioned the builder pattern in your second comment. But in case it's not that kind of pattern, it has to be all in a single argument, and it can't be a type. So it probably has to be something like `LocalTransport::<LocationSet!(Alice, Bob)>::channel()`

For the third point; yeah, it makes more sense to me too this way.

Reply:
My point was to move the constructor of `LocalTransportChannel` to `LocalTransport` and you're right that we need an explicit type parameter for the location set. 

The reason I prefer the builder pattern is that users don't have to remember `LocationSet!`. If we use `LocalTransport::<LocationSet!(Alice, Bob)>::channel()`, then what will happen is that

1. Users try to create a channel and just call `LocalTransport::channel()` and get an error saying they need to specify the non-inferrable type
2. They can go as far as `LocalTransport::<???>::channel()` but they cannot fill the question marks
3. Read the documentation, import `LocationSet!` and fix the issue

whereas if we use the builder pattern, users can simply follow the auto-complete.

Reply:
Ok; I see your point. I'll try doing that